Allow never-type goals in lockout

RasmusAntons opened the pull request:
After transitioning from a different mod for lockout, we wanted to be able to reimplement some of the goals in a datapack, especially negative goals that prohibit you from doing something in the game.

This code probably isn't great, but it adds a possible implementation for that. The logic is that initially every team is eligible to get the goal, shown by a split background colour. Once all but one team have lost the condition, that last team gets the point.

![Screenshot_20240512_212319](https://github.com/Gaming32/bingo/assets/6364347/3a6f4d4c-2556-47fe-a3ef-e6668008f94e)


This probably shouldn't be active by default because it requires some consideration for goals. The goals need to either be easy to get accidentally, like "never take fall damage" or be possible for other teams to enforce, like "never die" (especially when playing with a [player tracker](https://github.com/RasmusAntons/playertracker)). I added the start parameter `--allow-never-goals-in-lockout` for that.

Gaming32 commented on common/src/main/java/io/github/gaming32/bingo/game/BingoGameMode.java:
```diff
@@ -168,12 +170,21 @@ class TeamValue {
-            return goal.goal().getTags().stream().allMatch(g -> g.tag().specialType() == BingoTag.SpecialType.NONE);
+        public boolean isGoalAllowed(BingoGoal.Holder goal, boolean allowNeverGoalsInLockout) {
+            return goal.goal().getTags().stream().allMatch((g) -> {
+                if (g.tag().specialType() == BingoTag.SpecialType.NONE)
+                    return true;
+                if (g.tag().specialType() == BingoTag.SpecialType.NEVER)
+                    return allowNeverGoalsInLockout;
```
Please use curly braces on non-trivial if bodies.

Gaming32 reviewed (changes requested):
Mostly just small things. However, I have another issue I'd like to bring up. If the last goal is a never goal, nothing special is done. It should be made such that the game immediately ends in this case. The logic for immediately ending a lockout is already there, so you just need to make it run that in this case.

Gaming32 reviewed (approved):
Looks fine. Still needs a big test, obviously.

Gaming32 reviewed (changes requested):
Oh, I forgot. There still needs to be a check for if all the remaining goals are never goals, and triggering an immediate end if so.

RasmusAntons commented:
> Oh, I forgot. There still needs to be a check for if all the remaining goals are never goals, and triggering an immediate end if so.

Would it be an option to make goals that will be skipped if nothing else is left another special tag? Some never goals would be okay at the end of the game, for example a game tied on 12-12 with "never die" left (if you're playing with a player tracker and are able to find and attack the other team).

RasmusAntons commented:
During playtesting today we also found out that a negative goal being awarded doesn't trigger the check for game winning so I'll work on that too.

RasmusAntons commented:
I think this works, but I'm not sure if the stalemate tag could cause confusion, being only relevant for lockout gamemode.

Gaming32 commented:
I'd prefer if the `stalemate` tag was the default case, since most `never` goals will lead to a stalemate if lost.

RasmusAntons committed:
implement never type goals for multiple teams

# Conflicts:
#	common/src/main/generated/.cache/fd92b0e9e07895138b66a79af3d38ddbb4f5c5ed
#	src/main/generated/data/bingo/bingo/tag/lockout_inflictable.json
#	src/main/java/io/github/gaming32/bingo/client/BingoClient.java
#	src/main/java/io/github/gaming32/bingo/game/BingoGame.java
#	src/main/java/io/github/gaming32/bingo/game/mode/LockoutGameMode.java

RasmusAntons committed:
fix wrong translation key

RasmusAntons commented:
> I'd prefer if the `stalemate` tag was the default case, since most `never` goals will lead to a stalemate if lost.

I replaced the stalemate tag with a `lockout_inflictable` tag that's on goals that shouldn't end the game now. Also tested the 1.21.4 version and as far as I can tell everything works.
